### PR TITLE
[v2.0.0] Implement scout loading pattern

### PR DESCRIPTION
**Pull request by phamann**
**⚠️ Note this is a PR into the v.2.0.0 branch and not master**

### TL;DR
Reimplements the [scout file](https://alexsexton.com/blog/2013/03/deploying-javascript-applications/) loading pattern to `index.ts` which is very similar to what we have in v1. We test for browser feature support and if it passes, we invoke the `init()` method which loads the WebWorker main library when the browser is ready.

### Notes
- I've tried to abstract the main buisness logic utility functions out of the index file into their own modules to keep it clean and aid testability
- We need the feature detection logic to work on all browsers, hence the util and not simply using `Object.hasOwnProperty()`
- Had to do some pretty gnarly mocking and type casting to be able to test the `init()` method, as we don't run webpack for unit tests (quite rightly) which means the `workerize-loader` doesn't transform the `./worker` module into a class. But it works and improves test coverage :tada:

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
+import { hasProperties } from "./util/object";
+import loadWhenDocumentReady from "./util/loadWhenDocumentReady";
+import Worker from "workerize-loader!./worker";
 
-if ("Worker" in window && "Promise" in window) {
-  const worker = (lib as any)() as typeof lib;
+// List of required features browser features
+const requiredFeatures = ["Worker", "Promise", "fetch"];
+// Test whether browser has required feature support
+const hasFeatureSupport = hasProperties(window, requiredFeatures);
```
Comment (clight-fastly) [inline src/index.ts]: Do you expect to use `hasProperties` for other things in the future? Writing a util seems a little overkill if something like this works:

```
let has = false;
try {
has = window.Worker && window.Promise && window.fetch
}
```

File: src/@types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+interface Client {
```
Comment (clight-fastly) [inline src/@types/index.d.ts]: Why global types here (`d.ts`)? Importing types makes it a bit easier to refactor and globals don't clash as easy.

File: src/util/loadWhenDocumentReady.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Utility to load a function when the browsers ready state is complete
+export default function loadWhenReady(fn: () => void): void {
+  if (document.readyState !== "complete") {
```
Comment (clight-fastly) [inline src/util/loadWhenDocumentReady.ts]: style nit: flipping so truthy condition hits first reads better to me

File: src/util/loadWhenDocumentReady.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    document.addEventListener(
+      "readystatechange",
+      (): void => {
+        if (document.readyState === "complete") {
+          fn();
+        }
+      }
+    );
+  } else {
```
Comment (clight-fastly) [inline src/util/loadWhenDocumentReady.ts]: Don't think you need this `else` ?

File: src/util/object.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+// Utility function to test whether an object has own nested property
+// This is used instead of Object.prototype.hasOwnProperty() due to browser support
+// Accpets object path in dot notation form
+export function hasProperty(obj: any, propertyPath: string): boolean {
```
Comment (clight-fastly) [inline src/util/object.ts]: If you keep this util, I would think `hasOwnProperty` with a `try` before it seems more battle-hardened

**Comment by clight-fastly** [comment]
A few thoughts but no real blockers. Thanks!

File: src/index.ts (hunk 1 above)
Comment (phamann) [reply src/index.ts]: Good question. Yes, we will use the `hasProperties` util within the main library as we do some feature detection for the various `window.perfromance` sub APIs. 

We no longer use it here, but the other part of this util is that it accepts a object path as a string` i.e. `my.nested.object`, which is useful when testing the deeper performance apis mentioned above.

File: src/@types/index.d.ts (hunk 2 above)
Comment (phamann) [reply src/@types/index.d.ts]: This file does/will contain the interface that we export to the global `window.FASTLY` on the host page and if we ever expose as a module via npm to load into other applications. Therefore it made sense for it to be global.

File: src/util/object.ts (hunk 5 above)
Comment (phamann) [reply src/util/object.ts]: Fixed in https://github.com/fastly/insights.js/pull/28/commits/b3c32dfc575acd74087d48a6b6c969abca0091d2

File: src/util/loadWhenDocumentReady.ts (hunk 3 above)
Comment (phamann) [reply src/util/loadWhenDocumentReady.ts]: Fixed in: https://github.com/fastly/insights.js/pull/28/commits/2eabbe6176cbf0781efb2ba4c14a0d0ee3f5e8e4

**Comment by phamann** [comment]
Thanks for the reviews @clight-fastly and @psbanka, I've now addressed all feedback.
